[AArch64][compiler-rt] Add a function returning the current vector length

kmclaughlin-arm opened the pull request:
get_runtime_vl emits a cntd instruction if SVE is available at runtime,
otherwise it will return 0.

kmclaughlin-arm committed:
[AArch64][compiler-rt] Add a function returning the current vector length

get_runtime_vl emits a cntd instruction if SVE is available at runtime,
otherwise it will return 0.

kmclaughlin-arm committed:
- Run clang-format

efriedma-quic commented on compiler-rt/lib/builtins/aarch64/sme-abi-init.c:
```diff
@@ -50,3 +52,24 @@ __attribute__((constructor(90)))
+    __init_cpu_features();
+}
+
+__attribute__((target("sve"))) long emit_cntd(void) {
```
`static`?

efriedma-quic commented on compiler-rt/lib/builtins/aarch64/sme-abi-init.c:
```diff
@@ -50,3 +52,24 @@ __attribute__((constructor(90)))
+  return vl;
+}
+
+long get_runtime_vl(void) {
```
This probably should be prefixed with `__aarch64`

kmclaughlin-arm replied:
Hi @efriedma-quic, I've added this routine to the ABI document in https://github.com/ARM-software/abi-aa/pull/263 and renamed it to `__arm_get_current_vg`.

efriedma-quic commented on compiler-rt/lib/builtins/aarch64/sme-abi-init.c:
```diff
@@ -2,6 +2,8 @@
 // See https://llvm.org/LICENSE.txt for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
+#include "../cpu_model/aarch64.c"
```
`cpu_model/aarch64.c` isn't a header; including it seem wrong.

efriedma-quic reviewed (commented):
This function should be in the ABI document before we merge it, I think.

kmclaughlin-arm committed:
- Changes based on support routine added to the ABI document, see:
  https://github.com/ARM-software/abi-aa/pull/263
  - Renamed get_runtime_vl to __arm_get_current_vg
  - Also return VG if currently in streaming-mode

- Added static to get_aarch64_cpu_features

- Added aarch64.h and included in sme-abi-init.c

kmclaughlin-arm committed:
- Replaced use of __arm_in_streaming_mode from arm_sme.h with __arm_sme_state,
  to check if currently in streaming mode.

kmclaughlin-arm committed:
- Renamed get_features to __get_aarch64_features in aarch64.h

efriedma-quic commented on compiler-rt/lib/builtins/cpu_model/aarch64.h:
```diff
@@ -0,0 +1,90 @@
+
+long long __get_aarch64_features(void);
+
+void CONSTRUCTOR_ATTRIBUTE __init_cpu_features(void);
```
You shouldn't need CONSTRUCTOR_ATTRIBUTE on the declaration here, I think.

efriedma-quic commented on compiler-rt/lib/builtins/aarch64/sme-abi-init.c:
```diff
@@ -50,3 +52,35 @@ __attribute__((constructor(90)))
+extern struct SME_STATE __arm_sme_state(void) __arm_streaming_compatible;
+
+__attribute__((target("sve"))) long
+__arm_get_current_vg(void) __arm_streaming_compatible {
```
Maybe split this into a separate file, so code using the other stuff defined in this file doesn't pull in the constructor?  I assume __arm_get_current_vg won't be used in most configurations.

kmclaughlin-arm replied:
I've split this into a separate file and replaced the use of `has_sme()` with `__aarch64_has_sme_and_tpidr2_el0`.

efriedma-quic commented on compiler-rt/lib/builtins/cpu_model/aarch64.h:
```diff
@@ -0,0 +1,90 @@
+  FEAT_INIT      // Used as flag of features initialization completion
+};
+
+long long __get_aarch64_features(void);
```
Looking again... maybe we should just poke at __aarch64_cpu_features directly, instead of adding this function?  __aarch64_cpu_features is already part of the ABI (clang codegen knows about it).

kmclaughlin-arm replied:
Thanks, I've removed `__get_aarch64_features`. I'd missed that clang was also using `__aarch64_cpu_features`.

kmclaughlin-arm committed:
- Removed CONSTRUCTOR_ATTRIBUTE from __init_cpu_features declaration.

- Removed __get_aarch64_features.

- Split out changes for __arm_get_current_vg into a new file (sme-abi-vg.c)
  and removed from sme-abi-init.c.

efriedma-quic reviewed (approved):
LGTM

yozhu commented:
This change results in `compiler-rt` build failure:

```
compiler-rt/lib/builtins/aarch64/sme-abi-vg.c:27:5: error: call to undeclared function '__init_cpu_features'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    27 |     __init_cpu_features();
       |     ^
compiler-rt/lib/builtins/aarch64/sme-abi-vg.c:33:60: error: use of undeclared identifier 'FEAT_SVE'
    33 |   bool HasSVE = __aarch64_cpu_features.features & (1ULL << FEAT_SVE);
```

The new C file, `sme-abi-vg.c`, doesn't have `#if !defined(DISABLE_AARCH64_FMV)`.
